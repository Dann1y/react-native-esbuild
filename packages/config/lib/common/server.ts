import path from 'node:path';
import { LOCAL_CACHE_DIR } from './core';

export const ASSET_PATH = '/assets';

export const getDevServerAssetPath = (): string => {
  return path.resolve(process.cwd(), LOCAL_CACHE_DIR, 'assets');
};
