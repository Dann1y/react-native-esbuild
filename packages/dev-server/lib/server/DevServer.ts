import type { Server as HTTPServer } from 'node:http';
import type { ReactNativeEsbuildBundler } from '@react-native-esbuild/core';
import { DEFAULT_HOST, DEFAULT_PORT } from '../constants';
import type { DevServerOptions } from '../types';

export abstract class DevServer {
  protected initialized = false;
  protected server?: HTTPServer;
  protected bundler?: ReactNativeEsbuildBundler;
  protected devServerOptions: Required<DevServerOptions>;

  constructor(devServerOptions: DevServerOptions) {
    this.devServerOptions = {
      root: devServerOptions.root ?? process.cwd(),
      port: devServerOptions.port ?? DEFAULT_PORT,
      host: devServerOptions.host ?? DEFAULT_HOST,
    };
  }

  public abstract initialize(
    onPostSetup?: (bundler: ReactNativeEsbuildBundler) => void | Promise<void>,
  ): Promise<this>;

  public abstract listen(onListen?: () => void): Promise<HTTPServer>;
}
