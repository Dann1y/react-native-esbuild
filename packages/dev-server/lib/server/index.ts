import http, { type Server as HTTPServer } from 'node:http';
import { parse } from 'node:url';
import type { Server as WebSocketServer } from 'ws';
import {
  createDevServerMiddleware,
  indexPageMiddleware,
} from '@react-native-community/cli-server-api';
import { ReactNativeEsbuildBundler } from '@react-native-esbuild/core';
import {
  createHotReloadMiddleware,
  createServeAssetMiddleware,
  createServeBundleMiddleware,
  createSymbolicateMiddleware,
} from '../middlewares';
import { toSafetyMiddleware } from '../helpers';
import { logger } from '../shared';
import { DEFAULT_PORT, DEFAULT_HOST } from '../constants';
import type { DevServerMiddlewareContext, DevServerOptions } from '../types';

export class ReactNativeEsbuildDevServer {
  private initialized = false;
  private bundler?: ReactNativeEsbuildBundler;
  private server?: HTTPServer;
  private devServerOptions: Required<DevServerOptions>;

  constructor(devServerOptions: DevServerOptions) {
    this.devServerOptions = {
      port: devServerOptions.port ?? DEFAULT_PORT,
      host: devServerOptions.host ?? DEFAULT_HOST,
    };
  }

  private assertBundler(
    bundler?: ReactNativeEsbuildBundler,
  ): asserts bundler is ReactNativeEsbuildBundler {
    if (!bundler) {
      throw new Error('bundler is not ready');
    }
  }

  private assertHTTPServer(
    httpServer?: HTTPServer,
  ): asserts httpServer is HTTPServer {
    if (!httpServer) {
      throw new Error('http server is not ready');
    }
  }

  initialize(): {
    server: ReactNativeEsbuildDevServer;
    bundler: ReactNativeEsbuildBundler;
  } {
    if (this.initialized && this.bundler) {
      logger.warn('dev server already initialized');
      return { server: this, bundler: this.bundler };
    }

    logger.debug('setup dev server', this.devServerOptions);
    logger.debug('create bundler instance');

    const { server: hotReloadServer, ...hr } = createHotReloadMiddleware();
    const { middleware, websocketEndpoints } = createDevServerMiddleware({
      port: this.devServerOptions.port,
      host: this.devServerOptions.host,
      watchFolders: [],
    });

    this.bundler = new ReactNativeEsbuildBundler();
    this.bundler.addListener('build-start', hr.updateStart);
    this.bundler.addListener('build-end', ({ revisionId }) => {
      hr.hotReload(revisionId);
      hr.updateDone();
    });

    const context: DevServerMiddlewareContext = {
      bundler: this.bundler,
      devServerOptions: this.devServerOptions,
    };

    logger.debug('setup middlewares');

    // middleware for http logging
    middleware.use((request, _response, next) => {
      if (request.method && request.url) {
        logger.debug(`[${request.method}] ${request.url}`, request.headers);
      }
      next();
    });

    [
      createServeAssetMiddleware(context),
      createServeBundleMiddleware(context),
      createSymbolicateMiddleware(context),
      indexPageMiddleware,
    ].forEach((m) => middleware.use(toSafetyMiddleware(m)));

    logger.debug('create http server');
    this.server = http.createServer(middleware);

    logger.debug('setup web socket');
    this.server.on('upgrade', (request, socket, head) => {
      if (!request.url) return;

      const { pathname } = parse(request.url);
      const handler = pathname
        ? (websocketEndpoints[pathname] as WebSocketServer)
        : null;

      /**
       * @see {@link https://github.com/facebook/metro/blob/v0.77.0/packages/metro/src/index.flow.js#L230-L239}
       */
      if (pathname === '/hot') {
        logger.debug('HMR is not supported');
        hotReloadServer.handleUpgrade(request, socket, head, (client) => {
          hotReloadServer.emit('connection', client, request);
        });
      } else if (handler) {
        handler.handleUpgrade(request, socket, head, (client) => {
          handler.emit('connection', client, request);
        });
      } else {
        socket.destroy();
      }
    });

    return { server: this, bundler: this.bundler };
  }

  listen(): HTTPServer {
    this.assertBundler(this.bundler);
    this.assertHTTPServer(this.server);
    const { host, port } = this.devServerOptions;

    return this.server.listen(port, () => {
      process.stdout.write('\n');
      logger.info(`dev server listening on http://${host}:${port}`);
      process.stdout.write('\n');
    });
  }
}
