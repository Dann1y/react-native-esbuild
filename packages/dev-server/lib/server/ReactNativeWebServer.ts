import http, {
  type Server as HTTPServer,
  type ServerResponse,
  type IncomingMessage,
} from 'node:http';
import type { ServeResult } from 'esbuild';
import { ReactNativeEsbuildBundler } from '@react-native-esbuild/core';
import {
  combineWithDefaultBundleOptions,
  type BundleOptions,
} from '@react-native-esbuild/config';
import { logger } from '../shared';
import type { DevServerOptions } from '../types';
import { createSymbolicateMiddleware } from '../middlewares';
import { DevServer } from './DevServer';

/**
 * development server for web platform
 */
export class ReactNativeWebServer extends DevServer {
  private bundleOptions: BundleOptions;
  private serveResult?: ServeResult;

  constructor(
    devServerOptions: DevServerOptions,
    bundleOptions?: Partial<BundleOptions>,
  ) {
    super(devServerOptions);
    this.bundleOptions = combineWithDefaultBundleOptions(bundleOptions ?? {});
    this.initialize();
  }

  private proxyHandler(
    request: IncomingMessage,
    response: ServerResponse,
  ): void {
    if (!this.serveResult) {
      logger.debug('esbuild context is not serving now');
      return;
    }

    const options = {
      hostname: this.serveResult.host,
      port: this.serveResult.port,
      path: request.url,
      method: request.method,
      headers: request.headers,
    };

    const proxyRequest = http.request(options, (proxyResponse) => {
      logger.debug('proxy request', options);

      if (typeof proxyResponse.statusCode !== 'number') {
        logger.warn('invalid response status', proxyResponse.statusCode);
        response.writeHead(404, { 'Content-Type': 'text/html' });
        response.end('<h1>internal error</h1>');
        return;
      }

      if (proxyResponse.statusCode === 404) {
        response.writeHead(404, { 'Content-Type': 'text/html' });
        response.end('<h1>not found</h1>');
        return;
      }

      response.writeHead(proxyResponse.statusCode, proxyResponse.headers);
      proxyResponse.pipe(response, { end: true });
    });

    request.pipe(proxyRequest, { end: true });
  }

  private parseBody(request: IncomingMessage): Promise<void> {
    const body: Buffer[] = [];
    return new Promise((resolve, _reject) => {
      request
        .on('data', (chunk: Buffer) => body.push(chunk))
        .on('end', () => {
          request.rawBody = Buffer.concat(body).toString();
          resolve();
        });
    });
  }

  initialize(): void {
    if (this.initialized) {
      logger.warn('dev server already initialized');
      return;
    }

    logger.debug('create http server');
    const bundler = new ReactNativeEsbuildBundler(this.devServerOptions.root);
    this.bundler = bundler;

    const symbolicateMiddleware = createSymbolicateMiddleware(
      { bundler, devServerOptions: this.devServerOptions },
      { webBundleOptions: this.bundleOptions },
    );

    this.server = http.createServer((request, response) => {
      this.parseBody(request).then(() => {
        // 1. send request to middleware
        // 2. send request to proxy handler if not handled
        symbolicateMiddleware(request, response, () => {
          this.proxyHandler(request, response);
        });
      });
    });
    this.initialized = true;
  }

  async listen(onListen?: () => void): Promise<HTTPServer> {
    this.assertBundler(this.bundler);
    this.assertHTTPServer(this.server);

    const server = this.server;
    const { host, port } = this.devServerOptions;

    const serveResult = await this.bundler.serve(this.bundleOptions);
    this.serveResult = serveResult;

    server.listen(port, () => {
      logger.info(`dev server listening on http://${host}:${port}`);
      logger.debug(`proxy to esbuild ${serveResult.host}:${serveResult.port}`);
      process.stdout.write('\n');
      onListen?.();
    });

    return this.server;
  }
}
