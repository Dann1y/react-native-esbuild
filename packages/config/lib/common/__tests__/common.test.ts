import { getDevServerAssetPath } from '../server';

describe('getDevServerAssetPath', () => {
  it('should match snapshot', () => {
    expect(getDevServerAssetPath()).toMatchSnapshot();
  });
});
