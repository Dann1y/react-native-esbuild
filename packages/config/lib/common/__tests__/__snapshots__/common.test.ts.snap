// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBuildStatusCachePath should match snapshot 1`] = `"/root/.rne/build-status.json"`;

exports[`getDevServerAssetPath should match snapshot 1`] = `"/root/.rne/assets"`;
